syntax = "proto3";

package order;


message Order {
  string id = 1;
  repeated OrderItem items = 2;
  double total_amount = 3;
  string status = 4;
}

message OrderItem {
  string id = 1;
  string product_name = 2;
  int32 quantity = 3;
  double price = 4;
}


message CreateOrderRequest {
  repeated OrderItem items = 1;
  double total_amount = 2;
}

message CreateOrderResponse {
  Order order = 1;
}

message GetOrderRequest {
  string id = 1;
}

message GetOrderResponse {
  Order order = 1;
}

message UpdateOrderStatusRequest {
  string id = 1;
  string status = 2;
}

message UpdateOrderStatusResponse {
  Order order = 1;
}

message CancelOrderRequest {
  string id = 1;
}

message CancelOrderResponse {
  string message = 1;
}

message ListOrdersRequest {}

message ListOrdersResponse {
  repeated Order orders = 1;
}

message OrderStatusUpdate {
  string order_id = 1;
  string status = 2;
  int64 timestamp = 3;
}


service OrderService {
  rpc CreateOrder(CreateOrderRequest) returns (CreateOrderResponse) {}
  rpc GetOrder(GetOrderRequest) returns (GetOrderResponse) {}
  rpc UpdateOrderStatus(UpdateOrderStatusRequest) returns (UpdateOrderStatusResponse) {}
  rpc CancelOrder(CancelOrderRequest) returns (CancelOrderResponse) {}
  rpc ListOrders(ListOrdersRequest) returns (ListOrdersResponse) {}
  rpc OrderStatusStream(ListOrdersRequest) returns (stream OrderStatusUpdate) {}
}